modules/statistics/AnalyticsAdapter.js

```diff
@@ -1,130 +1,344 @@
+        const cache = this.cache;
+
+        this.cache = null;
+        if (cache) {
+            cache.forEach(event => this._sendEvent(event));
+        }
     }
 
     /**
-     * Cache analytics event.
-     * @param {String} eventName the name of the event
-     * @param {Object} data can be any JSON object
+     * Adds a set of permanent properties to add this this AnalyticsAdapter.
+     * Permanent properties will be added as "attributes" to events sent to
+     * the underlying "analytics handlers", and their keys will be prefixed
+     * by "permanent.", i.e. adding a permanent property {key: "value"} will
+     * result in {"permanent.key": "value"} object to be added to the
+     * "attributes" field of events.
+     *
+     * @param {Object} properties the properties to add
      */
-    sendEvent(eventName, data = {}) {
-        this.eventCache.push({
-            eventName,
-            data
-        });
+    addPermanentProperties(properties) {
+        for (const property in properties) {
+            if (properties.hasOwnProperty(property)) {
+                this.permanentProperties[`permanent.${property}`]
+                    = properties[property];
+            }
+        }
     }
 
     /**
-     * Clears the cached events.
-     * @returns {Array} with the cached events.
+     * Sends an event with a given name and given properties. The event type
+     * is set to "operational", and the required fields are all set to the given
+     * event name.
+     * @param {String|Object} eventName either a string to be used as the name
+     * of the event, or an event object. If an event object is passed, the
+     * properties parameters is ignored.
+     * @param {Object} properties the properties/attributes to attach to the
+     * event, if eventName is a string.
      */
-    drainCachedEvents() {
-        const eventCacheCopy = this.eventCache.slice();
+    sendEvent(eventName, properties = {}) {
+        let event = null;
+
+        if (typeof eventName === 'string') {
+            event = {
+                type: TYPE_OPERATIONAL,
+                action: eventName,
+                actionSubject: eventName,
+                source: eventName,
+                attributes: properties
+            };
+        } else if (typeof eventName === 'object') {
+            event = eventName;
+        }
 
-        this.eventCache = [];
+        if (!AnalyticsAdapter._verifyRequiredFields(event)) {
+            logger.warn(
+                `Dropping a mis-formatted event: ${JSON.stringify(event)}`);
 
-        return eventCacheCopy;
+            return;
+        }
+
+        this._sendEvent(event);
     }
 
-}
+    /**
+     *
+     * @param event
+     */
+    sendOperationalEvent(event) {
+        event.type = TYPE_OPERATIONAL;
 
-const cacheAnalytics = new CacheAnalytics();
+        if (!AnalyticsAdapter._verifyRequiredFields(event)) {
+            logger.warn(
```

Comment:
I like this. I wonder if it's even worth logging an error so it's more obvious when developing.

Reply:
I don't understand the difference in the way errors and warnings are shown, but I'll trust you on this and make it an error.